### PR TITLE
statistics: fix the top-n size to not hold the small things

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Problem Summary:

This pr prunes the not significant common items out of the top-n list.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

1. We are sampling without replacement, it's a hypergeometric distribution. We can calculate the variance and std derivation.
2. We assume that the element not in the top-n list shares the same occurrence. Thus the mean is the same.
3. We choose the continuity-corrected Wald interval as the most elements' occurrence.

### Check List 

Tests 

- Unit test(to be added later)
- Integration test on large datasets.

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: statistics/builder.go [hunk 1]
```diff
@@ -366,3 +368,55 @@ func BuildHistAndTopN(
 
 	return hg, topn, nil
 }
+
+// pruneTopNItem tries to prune the least common values in the top-n list if it is not significantly more common than the values not in the list.
+//   We assume that the ones not in the top-n list's selectivity is 1/remained_ndv which is the internal implementation of EqualRowCount
+func pruneTopNItem(topns []TopNMeta, ndv, nullCount, sampleRows, totalRows int64) []TopNMeta {
```
Comment (zhangjinpeng1987) [inline statistics/builder.go:374]: Is there any test for this function?

File: statistics/builder.go [hunk 2]
```diff
@@ -303,6 +304,7 @@ func BuildHistAndTopN(
 		}
 		cur, curCnt = sampleBytes, 1
 	}
+	topNList = pruneTopNItem(topNList, ndv, nullCount, sampleNum, count)
```
Comment (time-and-fate) [inline statistics/builder.go:307]: This should be after L332, there are logics of inserting the last value for TopN in L314 - L332.

File: statistics/builder.go (hunk 1 above)
Comment (winoros) [reply statistics/builder.go:375]: added

**Review by qw4990 — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]
lgtm

**Pull request merged**